ktor-server/ktor-server-plugins/ktor-server-rate-limit/jvmAndNix/src/io/ktor/server/plugins/ratelimit/RateLimit.kt

```diff
@@ -0,0 +1,199 @@
+    application.install(RateLimitApplicationInterceptors) {
+        this.providerNames = listOf(LIMITER_NAME_GLOBAL)
+    }
+}
+
+/**
+ * A config for [RateLimit] plugin.
+ */
+@KtorDsl
+public class RateLimitConfig {
+
+    internal val providers: MutableMap<String, RateLimitProvider> = mutableMapOf()
+    internal var global: RateLimitProvider? = null
+
+    /**
+     * Registers Rate-Limit provider that can be used in sub routes via [Route.rateLimit] function.
+     */
+    public fun register(name: String = LIMITER_NAME_EMPTY, block: RateLimitProviderConfig.() -> Unit) {
+        if (providers.containsKey(name)) {
+            throw IllegalStateException("Rate limit provider with name $name is already configured")
+        }
+        providers[name] = RateLimitProvider(
+            RateLimitProviderConfig().apply {
+                this.name = name
+                block()
+            }
+        )
+    }
+
+    /**
+     * Registers Rate-Limit provider that is applied to whole application
+     */
+    public fun global(block: RateLimitProviderConfig.() -> Unit) {
+        global = RateLimitProvider(
+            RateLimitProviderConfig().apply {
+                this.name = LIMITER_NAME_GLOBAL
+                block()
+            }
+        )
+    }
+}
+
+/**
+ * A config for [RateLimit] provider inside [RateLimiterConfig].
+ */
+@KtorDsl
+public class RateLimitProviderConfig {
+
+    internal var name: String = LIMITER_NAME_EMPTY
+    internal var requestKey: suspend (ApplicationCall) -> Any = { }
+    internal var requestWeight: suspend (ApplicationCall, key: Any) -> Int = { _, _ -> 1 }
+
+    internal var modifyResponse: (ApplicationCall, RateLimiter.State) -> Unit = { call, state ->
+        when (state) {
+            is RateLimiter.State.Available -> {
+                if (!call.response.headers.contains("X-RateLimit-Limit"))
+                    call.response.header("X-RateLimit-Limit", state.limit.toString())
+                if (!call.response.headers.contains("X-RateLimit-Remaining"))
+                    call.response.header("X-RateLimit-Remaining", state.remainingTokens.toString())
+                if (!call.response.headers.contains("X-RateLimit-Reset"))
+                    call.response.header("X-RateLimit-Reset", (state.refillAtTimeMillis / 1000).toString())
+            }
+            is RateLimiter.State.Exhausted -> {
+                if (!call.response.headers.contains(HttpHeaders.RetryAfter))
+                    call.response.header(HttpHeaders.RetryAfter, (state.timeToWaitMillis / 1000).toString())
+            }
+        }
+    }
+
+    internal var rateLimiterProvider: ((call: ApplicationCall, key: Any) -> RateLimiter)? = null
+    internal var rateLimiterConfig: RateLimiterConfig? = null
+
+    /**
+     * Sets [RateLimit] for this provider based on [ApplicationCall] and `key` of this request.
+     */
+    public fun rateLimiter(provider: (call: ApplicationCall, key: Any) -> RateLimiter) {
```

Comment:
what about an idea of introducing `fun interface`s for such lambda declarations?
It will be easier to instantiate such lambdas not in place, but externally and then passing here as parameter?